File: pherix/core/adapters/sql.py
"""SQLiteAdapter — the reversible adapter, correct by construction.

The database does the heavy lifting: ``snapshot`` issues a real ``SAVEPOINT``,
``restore`` does ``ROLLBACK TO SAVEPOINT``. The connection must be opened in
autocommit mode (``isolation_level=None``) so this adapter — not sqlite3's
implicit machinery — controls every BEGIN / SAVEPOINT / COMMIT / ROLLBACK.

The :class:`SQLiteAdapter` class never imports ``core/tools.py`` at module
load: the runtime resolves ``effect.tool`` to a callable and passes it to
``apply`` as ``tool_fn``. Slice 4's :func:`execute_isolated` helper does
read the ``active_effect`` contextvar from ``core/tools.py``, but only via
a function-local import to keep module-load ordering free of cycles.
"""

from __future__ import annotations

import json
import sqlite3
from typing import Any, Callable

from pherix.core.adapters.base import SnapshotHandle
from pherix.core.effects import Effect

# Side-table holding monotonic version counters per (resource, key).
# ``key_json`` is ``json.dumps(list(key), sort_keys=True)`` so cross-process
# readers (multiple Python processes against the same SQLite file) see
# consistent rows. The table is created on adapter init.
_VERSIONS_TABLE_DDL = """
CREATE TABLE IF NOT EXISTS _pherix_versions (
    resource TEXT NOT NULL,
    key_json TEXT NOT NULL,
    version  INTEGER NOT NULL DEFAULT 0,
    PRIMARY KEY (resource, key_json)
)
"""

# Seam used by the module-level :func:`execute_isolated` helper to find the
# :class:`SQLiteAdapter` bound to a given connection. We cannot attach
# attributes to a C-extension ``sqlite3.Connection`` directly (it forbids
# both attribute-set and weakref), so we key by ``id(conn)`` and require
# every adapter to register here on construction. The mapping leaks an
# entry per Connection object created in-process — a non-issue in test /
# agent-runtime contexts where adapters and connections are O(handful).
# A future revision could ship an explicit ``unregister`` if a long-lived
# process churns many connections.
_CONN_ADAPTERS: dict[int, "SQLiteAdapter"] = {}


def _adapter_for(conn: sqlite3.Connection) -> "SQLiteAdapter | None":
    """Return the :class:`SQLiteAdapter` registered for ``conn`` if any."""
    return _CONN_ADAPTERS.get(id(conn))


class SQLiteAdapter:
    name = "sql"

    def __init__(self, conn: sqlite3.Connection):
        self._conn = conn
        # Slice 4: create the version side-table eagerly so the first
        # read_version on an unknown key returns 0 (not a missing-table
        # error). DDL is idempotent so re-binding the adapter is safe.
        self._conn.execute(_VERSIONS_TABLE_DDL)
        # Slice 4 seam: register this adapter against the connection's
        # identity so the module-level :func:`execute_isolated` helper can
        # find it from a bare ``sqlite3.Connection`` (SQL tools accept a
        # plain Connection — no adapter parameter pollutes the call-site).
        # We must use an id-keyed module dict because ``sqlite3.Connection``
        # is a C-extension type that forbids both attribute assignment and
        # weakrefs.
        _CONN_ADAPTERS[id(conn)] = self
        # Slice 4 (D5 multi-process arbitration): a separate autocommit
        # "meta" connection used only for ``read_version`` queries.
        # Rationale: during an ``agent_txn`` the main connection sits
        # inside an open ``BEGIN``. Under WAL mode (and even some non-WAL
        # cases) its reads of ``_pherix_versions`` are snapshot-isolated
        # to the moment the txn started — so a bump committed by another
        # process *while we are open* would be invisible to us, and the
        # commit-time diff would silently miss the cross-process
        # lost-update. The meta-connection is never inside a BEGIN; its
        # reads see the latest committed state. Writes still go through
        # the main connection so they roll back with the txn.
        # In-memory DBs (``:memory:``) have no shareable path; meta_conn
        # is None and ``read_version`` falls back to the main conn —
        # which is fine, in-memory is single-process by definition.
        db_path = self._derive_db_path(conn)
        self._meta_conn: sqlite3.Connection | None = (
            sqlite3.connect(db_path, isolation_level=None)
            if db_path
            else None
        )

    @staticmethod
    def _derive_db_path(conn: sqlite3.Connection) -> str | None:
        """Path of the connection's ``main`` database, or None for memory.

        ``PRAGMA database_list`` yields rows ``(seq, name, file)`` — file
        is an empty string for ``:memory:`` connections. We honour only
        the ``main`` schema; attached schemas are out of scope (Slice 4
        does not promise multi-schema isolation).
        """
        for _seq, name, file in conn.execute("PRAGMA database_list").fetchall():
            if name == "main":
                return file or None
        return None

    @property
    def conn(self) -> sqlite3.Connection:
        return self._conn

    def supports_rollback(self) -> bool:
        return True

    # --- transaction-scope lifecycle (driven by the runtime, D3) ---

    def begin(self) -> None:
        self._conn.execute("BEGIN")

    def commit(self) -> None:
        self._conn.execute("COMMIT")

    def rollback(self) -> None:
        self._conn.execute("ROLLBACK")

    # --- per-effect snapshot / apply / restore ---

    @staticmethod
    def _savepoint_name(index: int) -> str:
        # SQLite cannot parameterise identifiers. `index` is an internal,
        # runtime-assigned integer (never user input), so building the
        # savepoint name by interpolation is safe by construction.
        return f"sp_{int(index)}"

    def snapshot(self, effect: Effect) -> SnapshotHandle:
        sp = self._savepoint_name(effect.index)
        self._conn.execute(f"SAVEPOINT {sp}")
        return SnapshotHandle(
            resource=self.name,
            effect_index=effect.index,
            payload={"savepoint": sp},
        )

    def apply(self, effect: Effect, tool_fn: Callable[..., Any]) -> Any:
        # D2: the txn-owned connection is injected as the tool's first arg;
        # the @tool wrapper hides it from the agent's call-site.
        return tool_fn(self._conn, **effect.args)

    def restore(self, handle: SnapshotHandle) -> None:
        sp = handle.payload["savepoint"]
        self._conn.execute(f"ROLLBACK TO SAVEPOINT {sp}")

    # --- versioning (Slice 4 — VersionedResourceAdapter) -------------------

    @staticmethod
    def _encode_key(key: tuple) -> str:
        # Pre-coercing the tuple to a list gives a stable JSON form across
        # processes regardless of how json handles tuples. ``sort_keys`` is
        # a no-op for lists but keeps the encoding canonical if a key ever
        # contains a nested dict.
        return json.dumps(list(key), sort_keys=True)

    def read_version(self, key: tuple) -> int:
        # Use the meta-connection if available (D5: it bypasses the main
        # connection's BEGIN snapshot, so cross-process bumps are visible
        # at commit-time diff). Falls back to the main connection for
        # in-memory adapters where meta_conn does not exist.
        target = self._meta_conn or self._conn
        row = target.execute(
            "SELECT version FROM _pherix_versions "
            "WHERE resource = ? AND key_json = ?",
            (self.name, self._encode_key(key)),
        ).fetchone()
        # Absent row → version 0 ("never written"). Never returns None.
        return 0 if row is None else int(row[0])

    def write_version(self, key: tuple) -> int:
        # Atomic UPSERT with RETURNING so the bump is race-free against a
        # second connection to the same on-disk SQLite file.
        cur = self._conn.execute(
            "INSERT INTO _pherix_versions (resource, key_json, version) "
            "VALUES (?, ?, 1) "
            "ON CONFLICT(resource, key_json) DO UPDATE "
            "SET version = version + 1 "
            "RETURNING version",
            (self.name, self._encode_key(key)),
        )
        row = cur.fetchone()
        return int(row[0])


# --- Slice 4 isolation helper ---------------------------------------------


def execute_isolated(
    conn: sqlite3.Connection,
    stmt: str,
    params: tuple = (),
    reads: list[tuple] | None = None,
    writes: list[tuple] | None = None,
) -> sqlite3.Cursor:
    """SQL execution that records read/write keys into the active Effect.

    ``reads`` and ``writes`` are lists of ``(table, pk_value)`` tuples — the
    keys this statement reads from and writes to. SQL parsing is out of
    scope, so Slice 4 ships explicit key declaration: tools say which rows
    they touched. A future iteration could derive the keys from the
    statement itself; the journal shape is the same either way.

    Behaviour:

    - Always runs ``conn.execute(stmt, params)`` and returns the cursor.
    - Inside an ``agent_txn``, records each read as ``("sql", tuple(key),
      adapter.read_version(key))`` into ``active_effect.read_keys`` and
      each write as ``("sql", tuple(key))`` into ``write_keys``, after
      bumping the side-table version via ``adapter.write_version(key)``.
      Re-recording the same triple / pair within one effect is suppressed
      so a tool that touches the same row repeatedly doesn't bloat the
      journal. The version-side-table bump still fires on every write
      call — every statement is a real write that increments the version,
      even if its presence in ``write_keys`` is already noted.
    - Outside an ``agent_txn`` (``active_effect`` is ``None``), the stmt
      still runs but no recording happens — keeps the helper usable from
      raw unit tests of SQL tools.
    - If the connection has no registered adapter (a bare
      ``sqlite3.Connection`` not wrapped by :class:`SQLiteAdapter`),
      recording is also skipped. This is the same graceful-degrade as
      the no-effect case.
    """
    # Local import to keep ``sql.py`` module-load free of any dependency on
    # ``core/tools.py`` (and the contextvar living there).
    from pherix.core.tools import active_effect

    cursor = conn.execute(stmt, params or ())
    effect = active_effect.get()
    if effect is None:
        return cursor
    adapter = _adapter_for(conn)
    if adapter is None:
        return cursor
    for key in reads or ():
        key_t = tuple(key)
        v = adapter.read_version(key_t)
        triple = ("sql", key_t, v)
        if triple not in effect.read_keys:
            effect.read_keys.append(triple)
    for key in writes or ():
        key_t = tuple(key)
        # Slice 4 P3: write_keys carries `(resource, key, version_after_my_write)`
        # so the commit-time diff can disambiguate self-bumps from cross-txn
        # writes via `last_my_write` lookup. The third element is "my expected
        # current" for this key after my write — if the live version differs,
        # someone else also wrote this key during my txn. Writes are NOT
        # deduplicated: repeated writes append fresh triples and the diff
        # picks the freshest via iteration order.
        v_after = adapter.write_version(key_t)
        effect.write_keys.append(("sql", key_t, v_after))
    return cursor

File: pherix-ts/src/adapters/sql.ts
/**
 * SqliteAdapter — the reversible lane proved against a real backend.
 * Mirror of pherix/core/adapters/sql.py, over the `better-sqlite3` driver.
 *
 * `snapshot()` issues a real `SAVEPOINT`; `restore()` does `ROLLBACK TO
 * SAVEPOINT`. The database does the heavy lifting — rollback is correct by
 * construction, needing no knowledge of what the effect meant. The whole
 * transaction is bracketed by `begin()` / `commit()` / `rollback()`.
 *
 * The driver type is kept structural (a minimal interface) so this module does
 * not force `better-sqlite3`'s types on consumers and so tests can substitute a
 * compatible fake if ever needed. At runtime the real driver satisfies it.
 */

import Database from "better-sqlite3";
import { activeEffect } from "../tools.js";
import type { Effect, SnapshotHandle } from "../effects.js";
import type { StateDiffable, ToolFn, TransactionalResourceAdapter } from "./base.js";

/** The slice of the better-sqlite3 Database surface this adapter uses. */
export interface SqliteDatabase {
  exec(source: string): unknown;
  prepare(source: string): {
    run(...params: any[]): unknown;
    get(...params: any[]): unknown;
    all(...params: any[]): unknown[];
    reader?: boolean;
  };
}

// Version side-table — the isolation substrate (#8). One monotonic counter per
// (resource, key). Created idempotently so the first readVersion of an unknown
// key returns 0.
const VERSIONS_TABLE_DDL = `
CREATE TABLE IF NOT EXISTS _pherix_versions (
    resource TEXT NOT NULL,
    key_json TEXT NOT NULL,
    version  INTEGER NOT NULL DEFAULT 0,
    PRIMARY KEY (resource, key_json)
)`;

/** Maps a live connection back to the adapter that wraps it, so the free
 *  function `executeIsolated(conn, ...)` can reach readVersion/writeVersion
 *  without changing the SQL tool calling convention. Mirrors Python's
 *  `_adapter_for(conn)`. */
const ADAPTER_FOR = new WeakMap<object, SqliteAdapter>();

export class SqliteAdapter implements TransactionalResourceAdapter, StateDiffable {
  readonly name = "sql";
  private readonly db: SqliteDatabase;
  // A separate autocommit "meta" connection used ONLY for readVersion, opened
  // when the main connection is backed by an on-disk file; null for :memory:
  // (no shareable path) or a driver that can't answer `PRAGMA database_list`.
  // Mirror of pherix.core.adapters.sql.SQLiteAdapter._meta_conn.
  private readonly metaDb: SqliteDatabase | null;

  constructor(db: SqliteDatabase) {
    this.db = db;
    // Create the version side-table eagerly (idempotent) so the first
    // readVersion on an unknown key returns 0, not a missing-table error.
    this.db.exec(VERSIONS_TABLE_DDL);
    ADAPTER_FOR.set(db as object, this);
    // The DDL above is committed (autocommit, pre-BEGIN), so the meta
    // connection opened now sees `_pherix_versions` immediately.
    this.metaDb = SqliteAdapter.openMetaConnection(db);
  }

  /**
   * Open a committed-only "meta" connection for readVersion, mirroring Python's
   * `_meta_conn`. During an agentTxn the main connection sits inside an open
   * BEGIN, so its reads of `_pherix_versions` reflect its own uncommitted
   * version bumps (and, under WAL, a stale read snapshot). The meta connection
   * is never inside a BEGIN: its reads see the latest *committed* state, so the
   * commit-time diff can compare committed-base-at-read against committed-base-
   * now and (a) not flag my own read-then-write of a key as a conflict, (b)
   * still catch a cross-connection committed write to a key I only read. Writes
   * still go through the main connection so they roll back with the txn.
   *
   * Returns null for :memory: (no shareable path → single-connection by
   * definition) and for any driver whose `PRAGMA database_list` we cannot read
   * (e.g. a test fake) — both keep the own-write-visible main-connection path.
   */
  private static openMetaConnection(db: SqliteDatabase): SqliteDatabase | null {
    const path = SqliteAdapter.derivePath(db);
    if (path === null) return null;
    // Autocommit (the better-sqlite3 default): the meta connection never opens
    // a transaction, so every readVersion sees the latest committed snapshot.
    // Like Python's, it is intentionally long-lived (one per on-disk adapter) —
    // a non-issue in agent-runtime contexts where adapters are O(handful).
    return new Database(path) as unknown as SqliteDatabase;
  }

  /** Path of the connection's `main` database, or null for :memory:. Mirrors
   *  Python's `_derive_db_path`: `PRAGMA database_list` yields (seq, name, file)
   *  rows; `file` is "" for an in-memory db. Honours only the `main` schema. */
  private static derivePath(db: SqliteDatabase): string | null {
    let rows: Array<{ name?: unknown; file?: unknown }>;
    try {
      rows = db.prepare("PRAGMA database_list").all() as Array<{ name?: unknown; file?: unknown }>;
    } catch {
      // A driver/fake that can't answer database_list → no committed-only path.
      return null;
    }
    for (const row of rows) {
      if (row.name === "main") {
        const file = typeof row.file === "string" ? row.file : "";
        return file === "" ? null : file;
      }
    }
    return null;
  }

  /** The underlying connection, for tools that need it directly in tests. */
  get connection(): SqliteDatabase {
    return this.db;
  }

  supportsRollback(): boolean {
    return true;
  }

  begin(): void {
    this.db.exec("BEGIN");
  }

  commit(): void {
    this.db.exec("COMMIT");
  }

  rollback(): void {
    this.db.exec("ROLLBACK");
  }

  private static savepointName(index: number): string {
    // index is internal (journal position), never user input — safe to inline.
    return `sp_${Math.trunc(index)}`;
  }

  snapshot(effect: Effect): SnapshotHandle {
    const sp = SqliteAdapter.savepointName(effect.index);
    this.db.exec(`SAVEPOINT ${sp}`);
    return { resource: this.name, effectIndex: effect.index, payload: { savepoint: sp } };
  }

  apply(effect: Effect, toolFn: ToolFn): unknown {
    // SQL tools receive the connection as their injected handle, then the
    // named-args object — the mirror of Python's tool_fn(conn, **args).
    return toolFn(this.db, effect.args);
  }

  restore(handle: SnapshotHandle): void {
    const sp = handle.payload["savepoint"] as string;
    // Rolls back to the savepoint, not beyond it: the parent txn stays open.
    this.db.exec(`ROLLBACK TO SAVEPOINT ${sp}`);
  }

  // --- versioning (isolation substrate, #8) --------------------------------

  private static encodeKey(key: unknown): string {
    // The key arrives as an array (e.g. ["accounts", "alice"]); JSON is its
    // canonical, stable encoding for the side-table primary key.
    return JSON.stringify(key);
  }

  /** Whether readVersion reflects ONLY committed state (excludes this txn's own
   *  uncommitted writes). True on-disk (the meta connection sits outside the
   *  main BEGIN); false for :memory: (single connection — reads see our own
   *  bumps, so the diff uses the own-write-visible branch). Mirror of Python's
   *  `reads_committed_only` → `self._meta_conn is not None`. */
  readsCommittedOnly(): boolean {
    return this.metaDb !== null;
  }

  /** Current version of `key`. Absent row → 0 ("never written"); never null.
   *  Reads through the committed-only meta connection when on-disk — so a
   *  cross-connection committed bump is visible at the commit-time diff and our
   *  own uncommitted bump is not — falling back to the main connection for
   *  :memory:, where no meta connection exists. */
  readVersion(key: unknown): number {
    const target = this.metaDb ?? this.db;
    const row = target
      .prepare("SELECT version FROM _pherix_versions WHERE resource = ? AND key_json = ?")
      .get(this.name, SqliteAdapter.encodeKey(key)) as { version: number } | undefined;
    return row === undefined ? 0 : Number(row.version);
  }

  /** Atomically bump `key`'s version; return the new value. */
  writeVersion(key: unknown): number {
    const row = this.db
      .prepare(
        "INSERT INTO _pherix_versions (resource, key_json, version) VALUES (?, ?, 1) " +
          "ON CONFLICT(resource, key_json) DO UPDATE SET version = version + 1 RETURNING version",
      )
      .get(this.name, SqliteAdapter.encodeKey(key)) as { version: number };
    return Number(row.version);
  }

  // --- state diff (StateDiffable) — for dry-run preview --------------------

  /** Names of user-created tables, excluding SQLite's own internal catalogue
   *  and Pherix's version side-table (bookkeeping, not user state). */
  private userTables(): string[] {
    const rows = this.db
      .prepare(
        "SELECT name FROM sqlite_master WHERE type = 'table' " +
          "AND name NOT LIKE 'sqlite_%' AND name != '_pherix_versions'",
      )
      .all() as Array<{ name: string }>;
    return rows.map((r) => r.name);
  }

  /** `{rowid: rowJson}` for one table, keyed by the stable implicit rowid so
   *  update-vs-insert can be told apart without parsing the schema. `table`
   *  comes from sqlite_master, never user input, so interpolating it is safe by
   *  construction (SQLite cannot parameterise identifiers regardless). */
  private dumpTable(table: string): Record<number, string> {
    const rows = this.db.prepare(`SELECT rowid AS __rowid, * FROM ${table}`).all() as Array<
      Record<string, unknown>
    >;
    const out: Record<number, string> = {};
    for (const row of rows) {
      const { __rowid, ...rest } = row;
      out[__rowid as number] = JSON.stringify(rest);
    }
    return out;
  }

  stateBaseline(): Record<string, Record<number, string>> {
    const out: Record<string, Record<number, string>> = {};
    for (const t of this.userTables()) out[t] = this.dumpTable(t);
    return out;
  }

  stateDiff(baseline: unknown): Record<string, unknown> {
    const base = baseline as Record<string, Record<number, string>>;
    const added: Array<{ table: string; row: unknown }> = [];
    const modified: Array<{ table: string; row: unknown }> = [];
    const deleted: Array<{ table: string; row: unknown }> = [];
    const liveTables = this.userTables();
    for (const table of liveTables) {
      const before = base[table] ?? {};
      const now = this.dumpTable(table);
      for (const [rowid, rowJson] of Object.entries(now)) {
        if (!(rowid in before)) added.push({ table, row: JSON.parse(rowJson) });
        else if (before[rowid as unknown as number] !== rowJson) {
          modified.push({ table, row: JSON.parse(rowJson) });
        }
      }
      for (const [rowid, rowJson] of Object.entries(before)) {
        if (!(rowid in now)) deleted.push({ table, row: JSON.parse(rowJson) });
      }
    }
    // A table present in the baseline but dropped during the txn: its rows
    // count as deletions so the diff stays honest about lost data.
    const liveSet = new Set(liveTables);
    for (const [table, before] of Object.entries(base)) {
      if (!liveSet.has(table)) {
        for (const rowJson of Object.values(before)) {
          deleted.push({ table, row: JSON.parse(rowJson) });
        }
      }
    }
    return { rows_added: added, rows_modified: modified, rows_deleted: deleted };
  }
}

/**
 * SQL execution that records read/write keys into the active Effect (#8).
 * Mirror of pherix.core.adapters.sql.execute_isolated.
 *
 * `reads` and `writes` are the keys this statement touches — each a tuple-shaped
 * array like `["accounts", "alice"]`. SQL parsing is out of scope, so tools
 * declare which rows they touched explicitly; the journal shape is the same as
 * if it were derived from the statement.
 *
 * Always runs the statement and returns its result (rows for a reader, the
 * run-info otherwise). Inside an agentTxn it records each read as
 * `["sql", key, readVersion(key)]` into the effect's readKeys (deduped) and each
 * write as `["sql", key, writeVersion(key)]` into writeKeys (after bumping the
 * side-table version — the version-after-my-write the commit diff compares
 * against). Outside an agentTxn (no active effect) or for a bare connection not
 * wrapped by a SqliteAdapter, recording is skipped — the statement still runs.
 */
export function executeIsolated(
  conn: SqliteDatabase,
  stmt: string,
  params: unknown[] = [],
  opts: { reads?: unknown[]; writes?: unknown[] } = {},
): unknown {
  const prepared = conn.prepare(stmt);
  const result = prepared.reader ? prepared.all(...params) : prepared.run(...params);

  const effect = activeEffect.getStore();
  const adapter = ADAPTER_FOR.get(conn as object);
  if (effect === undefined || adapter === undefined) return result;

  const seenReads = new Set(effect.readKeys.map(([, k]) => JSON.stringify(k)));
  for (const key of opts.reads ?? []) {
    const enc = JSON.stringify(key);
    if (seenReads.has(enc)) continue;
    effect.readKeys.push(["sql", key, adapter.readVersion(key)]);
    seenReads.add(enc);
  }
  for (const key of opts.writes ?? []) {
    // write_keys carries (resource, key, versionAfterMyWrite) so the diff can
    // disambiguate self-bumps from cross-txn writes. Not deduped: repeated
    // writes append fresh triples; the diff picks the freshest by order.
    const vAfter = adapter.writeVersion(key);
    effect.writeKeys.push(["sql", key, vAfter]);
  }
  return result;
}
